C daemon becomes unresponsive

### Describe the bug

After being left running for some hours I'm unable to connect to a C sshnpd that was previously working.

This isn't an isolated occurrence, and has been seen on multiple installations. 👀 @cconstab and @gkc 

From a serial console I can see that the process is running:

```log
root@One:~# ps -www | grep sshnpd
 4203 root      2296 R    /usr/bin/sshnpd -a @bareindoornetball -m @cpswan -d openwrtone
 4754 root      2556 S    /usr/bin/sshnpd -a @bareindoornetball -m @cpswan -d openwrtone
```

And there's a connection to the device atServer:

```log
root@One:~# netstat -an | grep 15014
tcp        0      0 10.10.10.34:58654       104.198.230.74:15014    ESTABLISHED
tcp        0      0 10.10.10.34:50700       35.226.184.123:15014    CLOSE_WAIT
```

NB the `CLOSE_WAIT` shows that at some stage the daemon has reconnected to the atServer and got a different IP from the load balancer.

### Steps to reproduce

1. First I install and configure the latest release of C sshnpd (in the case using our [OpenWrt package for c0.2.4](https://github.com/atsign-foundation/Atsign_OpenWRT_packages/releases/tag/c0.2.4)
2. Then I test that everything working by doing some `sshnp` sessions
3. And then leave things for a few hours
4. Then try again to log in. `sshnp` times out at `Waiting for daemon feature check response`


### Expected behavior

`sshnp` connects

### Additional context

An ssh session established with the previously working sshnpd survived whatever is happening here. Closing it makes no difference.

Yay I saw the same before and now I have the issue again.. (All times in UTC)
The daemon is running at 100% CPU
```
    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND                                           17754 cconstab  20   0   78904   3072   2496 R  99.9   0.6     38,38 /home/cconstab/.local/bin/sshnpd -a @ssh_1 -m + 
 ```
 
 Not responding to pings.. And the Noop loop seems broken as I see no logs in the tmux.. Video to come.. this is the last of the log in TMUX

```
EBG] 2024-11-27 21:12:05.929943 | connection |        RECV: "data:218516"
[DEBG] 2024-11-27 21:12:05.929956 | refresh_device_entry | Released the atclient lock
[INFO] 2024-11-27 22:12:07.149668 | refresh_device_entry | Refreshing device info entries for this device
[DEBG] 2024-11-27 22:12:07.149764 | connection |        SENT: "noop:0"
[ERROR] 2024-11-27 22:12:07.227495 | connection | mbedtls_ssl_read failed with exit code: -80
[ERROR] 2024-11-27 22:12:07.227500 | connection | Failed to send "noop:0
" to connection: -80
[INFO] 2024-11-27 22:12:07.227505 | reconnect | Worker client is not connected, attempting to reconnect:
[INFO] 2024-11-27 22:12:07.227509 | atclient | Missing atServer host or port. Using production atDirectory to look up atServer host and port
[DEBG] 2024-11-27 22:12:07.678725 | connection |        SENT: "ssh_1"
[DEBG] 2024-11-27 22:12:07.746729 | connection |        RECV: "92a7c219-e2eb-5f50-ac23-d7acba106cf5.swarm0002.atsign.zone:5703"
[DEBG] 2024-11-27 22:12:08.239173 | connection |        SENT: "from:ssh_1"
[DEBG] 2024-11-27 22:12:08.306705 | connection |        RECV: "data:_a2ada53a-9f2c-41e7-9acf-a12d36ad48e4@ssh_1:da63d2b8-6ab3-4098-96d0-94d26c4c3a53"
[DEBG] 2024-11-27 22:12:08.324493 | connection |        SENT: "pkam:KUDMZuqAa/naesCStDxx75XQ0ne5AZDLNIuvZZ4aStyHJJrqIvG6hUqR/rk7LEmotD6Egbc237haWotVbYGGkfsGeh4C+ZbE4A93ZFxu5+tM1bDaUPNbiFoDBqORQJJB6gyxzzIpS8pdHU8RTdhchJN8j0dw8A+Jt/TGyw34VMcUy2PlxEaqonC6Ej0VtXHibXzBQLx/LBXjNVO2GoqzaBCFHzQULRFXVS1QxIz69Ax36RzSC0IjaPjIWRJ65+f4YAopSapWFzwMblOIda8s17Gh1eSuNOVfKFMEiCqLCyNvBXQB9zSJUMCXO/W0yyjroPob2A5NFD4oMHptx4n0vA=="
[DEBG] 2024-11-27 22:12:08.403668 | connection |        RECV: "data:success"
[INFO] 2024-11-27 22:12:08.403674 | reconnect | Reconnected to the atServer!
[DEBG] 2024-11-27 22:12:08.403690 | connection |        SENT: "llookup:shared_key.cconstab@ssh_1"
[DEBG] 2024-11-27 22:12:08.479672 | connection |        RECV: "data:k0Er2lWbM2bWWjlGDTII4b8xI8pg+++3DJ8Vouw7rPcKu8oXIsg+3bWCcMWOJ1VpqmshXN76Ry4OAENJwF8iXIqxC7UznDdCq0/cdCDD9UIQol7Dxwj5I+1fIzLbJNDWkey1LBkqcVxCzbOZvn323e4+Hua+UnFQPtU6TCkwKstlA62vT7WoInlDPRnwoOojABD7037N8DGFPeZjQAGnlQYK7+d9SZgQqIhgJhWAHX3TUUHpQRVPBmTBIWVZYbOsi+k7WRWVDIFP4rmu2DP5cJklBbSPCM2Pa8KkYszEfa5/HQGYcI8n6b3B3lyCbL/UT+iDl9ay6yY6epkHVL1fOw=="
[DEBG] 2024-11-27 22:12:08.493921 | connection |        SENT: "noop:0"
[DEBG] 2024-11-27 22:12:08.562817 | connection |        RECV: "data:ok"
[DEBG] 2024-11-27 22:12:08.562879 | connection |        SENT: "update:ttl:2592000000:ttr:-1:ccd:true:isEncrypted:true:ivNonce:t3+Pyv31/kVb2bfytXxg5w==:@cconstab:device_info.csshnp01.sshnp@ssh_1 JVMw3MIUZWBHZMHIHk6m7oeyfk0JR9cr4GTuNldOlXFuPTdwFsXVfipA4HzoNsL7yL+RRJuwAVnnwk338wPSmhz5AmccVDq5xq99WtEB3TWYPHj+S2tVFwulktLM/+HBHsc6xa1Lm32VVboeaxYs63nT2hg4bXbllVplpWzFwdu6xJg1AcmnQevszWsYnIz+SgPscRfTHTBAZtpNDMQvvr3tEBvF4LMHHuysMYkGxmOIcfp1q+tAVFQ3z3Rto2NeQVgsRf4NN703KHeYQ4rlUBrKOaKlmxD51ARGqzhk0LWcQlx8vrUFglm01Ugs/xOD"
[DEBG] 2024-11-27 22:12:08.659503 | connection |        RECV: "data:218549"
[DEBG] 2024-11-27 22:12:08.659509 | refresh_device_entry | Released the atclient lock
```
 
 Video

https://github.com/user-attachments/assets/fa95979b-99da-4f24-be2d-b1b1f88e612a

csshnpd log file

[csshnpd.log](https://github.com/user-attachments/files/17940714/csshnpd.log)

I also have logs...

From the last successful monitor:

```log

Wed Nov 27 17:01:19 2024 daemon.info sshnpd[10188]: [DEBG] 2024-11-27 17:01:19.474620 | sshnpd | Waiting for next monitor thread message
Wed Nov 27 17:01:34 2024 daemon.info sshnpd[10188]: [DEBG] 2024-11-27 17:01:34.474036 | atclient_monitor |      RECV: "notification: {"id":"-1","from":"@bareindoornetball","to":"@bareindoornetball","key":"statsNotification.@bareindoornetball","value":"78978","operation":"update","epochMillis":1732726894425,"messageType":"MessageType.key","isEncrypted":false,"metadata":null}"
Wed Nov 27 17:01:34 2024 daemon.info sshnpd[10188]: [DEBG] 2024-11-27 17:01:34.474167 | sshnpd | Received message of type: 1
Wed Nov 27 17:01:34 2024 daemon.info sshnpd[10188]: [DEBG] 2024-11-27 17:01:34.474189 | sshnpd | Notification value received: 78978
Wed Nov 27 17:01:34 2024 daemon.info sshnpd[10188]: [DEBG] 2024-11-27 17:01:34.474209 | sshnpd | Waiting for next monitor thread message
Wed Nov 27 17:48:12 2024 daemon.info sshnpd[10188]: [INFO] 2024-11-27 17:48:12.124339 | refresh_device_entry | Refreshing device info entries for this device
Wed Nov 27 17:48:12 2024 daemon.info sshnpd[10188]: [DEBG] 2024-11-27 17:48:12.124564 | connection |    SENT: "noop:0"
Wed Nov 27 17:48:12 2024 daemon.info sshnpd[10188]: [ERROR] 2024-11-27 17:48:12.124640 | connection | mbedtls_ssl_read failed with exit code: 0
Wed Nov 27 17:48:12 2024 daemon.info sshnpd[10188]: [DEBG] 2024-11-27 17:48:12.124676 | connection |    SENT: "llookup:shared_key.cpswan@bareindoornetball"
Wed Nov 27 17:48:12 2024 daemon.info sshnpd[10188]: [ERROR] 2024-11-27 17:48:12.124702 | connection | mbedtls_ssl_read failed with exit code: 0
Wed Nov 27 17:48:12 2024 daemon.info sshnpd[10188]: [ERROR] 2024-11-27 17:48:12.124720 | base64 | atchops_base64_decode: src_len is less than or equal to 0
Wed Nov 27 17:48:12 2024 daemon.info sshnpd[10188]: [ERROR] 2024-11-27 17:48:12.124737 | encryption_key_helpers | atchops_base64_decode: 1
Wed Nov 27 17:48:12 2024 daemon.info sshnpd[10188]: [ERROR] 2024-11-27 17:48:12.124754 | atclient_put_shared_key | atclient_get_shared_encryption_key_shared_by_me: 1
Wed Nov 27 17:48:12 2024 daemon.info sshnpd[10188]: [ERROR] 2024-11-27 17:48:12.124773 | refresh_device_entry | Failed to refresh device entry for @cpswan
Wed Nov 27 17:48:12 2024 daemon.info sshnpd[10188]: [DEBG] 2024-11-27 17:48:12.124790 | refresh_device_entry | Released the atclient lock
Wed Nov 27 18:48:13 2024 daemon.info sshnpd[10188]: [INFO] 2024-11-27 18:48:13.402860 | refresh_device_entry | Refreshing device info entries for this device
Wed Nov 27 18:48:13 2024 daemon.info sshnpd[10188]: [ERROR] 2024-11-27 18:48:13.402967 | connection | mbedtls_ssl_write failed with exit code: -80
Wed Nov 27 18:48:13 2024 daemon.info sshnpd[10188]: [ERROR] 2024-11-27 18:48:13.402987 | connection | Failed to send "noop:0
Wed Nov 27 18:48:13 2024 daemon.info sshnpd[10188]: " to connection: -80
Wed Nov 27 18:48:13 2024 daemon.info sshnpd[10188]: [INFO] 2024-11-27 18:48:13.403004 | reconnect | Worker client is not connected, attempting to reconnect:
Wed Nov 27 18:48:13 2024 daemon.info sshnpd[10188]: [INFO] 2024-11-27 18:48:13.403020 | atclient | Missing atServer host or port. Using production atDirectory to look up atServer host and port
Wed Nov 27 18:48:13 2024 daemon.info sshnpd[10188]: [DEBG] 2024-11-27 18:48:13.984917 | connection |    SENT: "bareindoornetball"
Wed Nov 27 18:48:14 2024 daemon.info sshnpd[10188]: [DEBG] 2024-11-27 18:48:14.086475 | connection |    RECV: "e6697d6a-2567-50a7-9e61-396157b733d8.swarm0001.atsign.zone:15014"
Wed Nov 27 18:48:14 2024 daemon.info sshnpd[10188]: [DEBG] 2024-11-27 18:48:14.662178 | connection |    SENT: "from:bareindoornetball"
Wed Nov 27 18:48:14 2024 daemon.info sshnpd[10188]: [DEBG] 2024-11-27 18:48:14.763326 | connection |    RECV: "data:_0400ed33-9583-43bc-8fbd-bb93e594eab6@bareindoornetball:8bf6e0ab-bfe0-4610-905d-07e5542aa021"
Wed Nov 27 18:48:14 2024 daemon.info sshnpd[10188]: [DEBG] 2024-11-27 18:48:14.805014 | connection |    SENT: "pkam:bjoNoPQsSexUP0WKcX3lm3xMs04AxVoD/ZeCDDkF3bO6jjto6egJA1+XiHPUyi4s7BhMYzn50D4Fnj6LRv6X/BF28f2a/SQL9b+psuEhU7UDoxjvs31JcOKA4JH6ZwwyNzC0OQQPolT1Ld/GZ2SyFXcd71LQFoC2WdZyBS8VZaXKW2T1TltSMnkgj3rJqO27knAJQJEIGfVRb9mChxKoSeY0XaYnH1G59UvQUYsa6Mh6PgjUIha3ulNrJ+NybwRafLx8XhsUOvqMvSjjPTsFSr737RQ2n5VaS4VP3KQFgSDIOzq4b9GlTMUK7F8kao6iyH9OroFbEaZa0yuGh73eeA=="
Wed Nov 27 18:48:14 2024 daemon.info sshnpd[10188]: [DEBG] 2024-11-27 18:48:14.906585 | connection |    RECV: "data:success"
Wed Nov 27 18:48:14 2024 daemon.info sshnpd[10188]: [INFO] 2024-11-27 18:48:14.907140 | reconnect | Reconnected to the atServer!
```

After that the monitor loop seems to fail

In the [full log](https://gist.github.com/cpswan/47632b39251221eebca2ce4db52349f3) I can see several (successful) attempts to reconnect to the atServer, but the monitor loop never seems to get re-established.

--- a/packages/c/cmake/atsdk.cmake
+++ b/packages/c/cmake/atsdk.cmake
@@ -3,7 +3,7 @@ if(NOT atsdk_FOUND)
   FetchContent_Declare(
     atsdk
     GIT_REPOSITORY https://github.com/atsign-foundation/at_c.git
-    GIT_TAG 8e48c9d542965a006e3bf66413177dd72205da72
+    GIT_TAG 9b9cb315a8af08a19e9679c9dad98572536f2ee4
   )
   FetchContent_MakeAvailable(atsdk)
   install(TARGETS atclient atchops atlogger)

--- a/packages/c/cmake/atsdk.cmake
+++ b/packages/c/cmake/atsdk.cmake
@@ -3,7 +3,7 @@ if(NOT atsdk_FOUND)
   FetchContent_Declare(
     atsdk
     GIT_REPOSITORY https://github.com/atsign-foundation/at_c.git
-    GIT_TAG 6650e33ab80e9d873ede6d0213fa2999054fce7b
+    GIT_TAG 8e48c9d542965a006e3bf66413177dd72205da72
   )
   FetchContent_MakeAvailable(atsdk)
   install(TARGETS atclient atchops atlogger)

--- a/packages/c/sshnpd/CHANGELOG.md
+++ b/packages/c/sshnpd/CHANGELOG.md
@@ -1,3 +1,7 @@
+## 0.2.5
+
+- fix: uptake some fixes in monitor
+
 ## 0.2.4
 
 - fix: Disabled clang-tidy missing-includes, as it malformed header includes

--- a/packages/c/sshnpd/include/sshnpd/version.h
+++ b/packages/c/sshnpd/include/sshnpd/version.h
@@ -1,4 +1,4 @@
 #ifndef SSHNPD_VERSION_H
 #define SSHNPD_VERSION_H
-#define SSHNPD_VERSION "0.2.4"
+#define SSHNPD_VERSION "0.2.5"
 #endif

--- a/packages/c/sshnpd/src/main.c
+++ b/packages/c/sshnpd/src/main.c
@@ -415,12 +415,16 @@ void main_loop() {
     // Read the next monitor message
     int ret = atclient_monitor_read(&monitor_ctx, &worker, &message, &monitor_hooks);
     if (ret != 0) {
-      atlogger_log(LOGGER_TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Possible bad state: monitor read failed\n");
+      atlogger_log(LOGGER_TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Possible bad state: monitor read failed (ret: %d)\n",
+                   ret);
     }
 
     atlogger_log(LOGGER_TAG, ATLOGGER_LOGGING_LEVEL_DEBUG, "Received message of type: %d\n", message.type);
 
     switch (message.type) {
+    case ATCLIENT_MONITOR_MESSAGE_TYPE_EMPTY:
+      // We got a timeout, nothing to read, nothing to do
+      break;
     case ATCLIENT_MONITOR_ERROR_READ:
       if (!atclient_monitor_is_connected(&monitor_ctx)) {
         atlogger_log(LOGGER_TAG, ATLOGGER_LOGGING_LEVEL_ERROR,

